spartan/aztec-network/templates/validator.yaml

```diff
@@ -179,62 +176,77 @@ data:
 ---
+# Headless service for StatefulSet DNS entries
 apiVersion: v1
 kind: Service
 metadata:
   name: {{ include "aztec-network.fullname" . }}-validator
   labels:
     {{- include "aztec-network.labels" . | nindent 4 }}
 spec:
-  type: ClusterIP
+  clusterIP: None
   selector:
     {{- include "aztec-network.selectorLabels" . | nindent 4 }}
     app: validator
   ports:
     - port: {{ .Values.validator.service.p2pTcpPort }}
       name: p2p-tcp
+      protocol: TCP
     - port: {{ .Values.validator.service.p2pUdpPort }}
       name: p2p-udp
       protocol: UDP
     - port: {{ .Values.validator.service.nodePort }}
       name: node
+      protocol: TCP
 ---
 {{if .Values.network.public }}
+{{- range $i, $e := until (int .Values.validator.replicas) }}
+# Service template for TCP load balancers
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ include "aztec-network.fullname" . }}-validator-lb-tcp
+  name: {{ include "aztec-network.fullname" $ }}-validator-{{ $i }}-lb-tcp
+  annotations:
+    service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
+    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: "ip"
+    service.beta.kubernetes.io/aws-load-balancer-scheme: "internet-facing"
   labels:
-    {{- include "aztec-network.labels" . | nindent 4 }}
+    {{- include "aztec-network.labels" $ | nindent 4 }}
 spec:
   type: LoadBalancer
   selector:
-    {{- include "aztec-network.selectorLabels" . | nindent 4 }}
+    statefulset.kubernetes.io/pod-name: {{ include "aztec-network.fullname" $ }}-validator-{{ $i }}
+    {{- include "aztec-network.selectorLabels" $ | nindent 4 }}
     app: validator
   ports:
-    - port: {{ .Values.validator.service.p2pTcpPort }}
+    - port: {{ $.Values.validator.service.p2pTcpPort }}
       name: p2p-tcp
-    - port: {{ .Values.validator.service.nodePort }}
+      protocol: TCP
+    - port: {{ $.Values.validator.service.nodePort }}
       name: node
+      protocol: TCP
 ---
+# Service template for UDP load balancers
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ include "aztec-network.fullname" . }}-validator-lb-udp
+  name: {{ include "aztec-network.fullname" $ }}-validator-{{ $i }}-lb-udp
   annotations:
     service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
     service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: "ip"
     service.beta.kubernetes.io/aws-load-balancer-scheme: "internet-facing"
   labels:
-    {{- include "aztec-network.labels" . | nindent 4 }}
+    {{- include "aztec-network.labels" $ | nindent 4 }}
 spec:
   type: LoadBalancer
   selector:
-    {{- include "aztec-network.selectorLabels" . | nindent 4 }}
+    statefulset.kubernetes.io/pod-name: {{ include "aztec-network.fullname" $ }}-validator-{{ $i }}
+    {{- include "aztec-network.selectorLabels" $ | nindent 4 }}
     app: validator
   ports:
-    - port: {{ .Values.validator.service.p2pUdpPort }}
+    - port: {{ $.Values.validator.service.p2pUdpPort }}
```

Comment:
Confused on this $. syntax too

Reply:
Without using the `$` global scope, the helm apply fails with: 
```
Error: template: aztec-network/templates/validator.yaml:222:22: executing "aztec-network/templates/validator.yaml" at <.Values.validator.service.p2pTcpPort>: can't evaluate field Values in type int
```